File: app/src/main/java/com/hussien/qouty/fragments/SettingsFragment.java
package com.hussien.qouty.fragments;

import android.app.AlertDialog;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.text.InputType;

import androidx.navigation.fragment.NavHostFragment;
import androidx.preference.EditTextPreference;
import androidx.preference.ListPreference;
import androidx.preference.MultiSelectListPreference;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;

import com.hussien.qouty.AppExecutor;
import com.hussien.qouty.R;
import com.hussien.qouty.data.QuotesRepository;
import com.hussien.qouty.utilities.AlarmUtil;
import com.mikepenz.aboutlibraries.LibsBuilder;


public class SettingsFragment extends PreferenceFragmentCompat implements SharedPreferences.OnSharedPreferenceChangeListener, Preference.OnPreferenceChangeListener {

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        setPreferencesFromResource(R.xml.preference, rootKey);

        EditTextPreference numberPreference = findPreference(getString(R.string.key_number_quotes));
        if (numberPreference != null){
            numberPreference.setOnBindEditTextListener(editText -> editText.setInputType(InputType.TYPE_CLASS_NUMBER));
            numberPreference.setOnPreferenceChangeListener(this);
        }

        MultiSelectListPreference tagsPreference = findPreference(getString(R.string.key_tags));
        if (tagsPreference != null) {
            tagsPreference.setEnabled(false);
            setTagsEntriesAndValues(tagsPreference);
        }

        checkNotificationEnabled();

        Preference aboutPreference = findPreference(getString(R.string.key_pref_about));
        if (aboutPreference != null){
            aboutPreference.setOnPreferenceClickListener(preference -> {
                navigateToAboutPage();
                return true;
            });
        }

        Preference creditsPreference = findPreference(getString(R.string.key_pref_credits));
        if (creditsPreference != null){
            creditsPreference.setOnPreferenceClickListener(preference -> {
                showCreditsDialog();
                return true;
            });
        }
    }

    private void showCreditsDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        builder.setTitle("Credits");
        builder.setMessage("Thanks to harbi for the quotes collection https://github.com/harbi/short-quotes\nApp Icon Designed by Freepik");
        builder.setPositiveButton("OK", (dialog, which) -> dialog.dismiss());
        builder.create().show();
    }

    private void navigateToAboutPage(){
        LibsBuilder builder = new LibsBuilder()
                .withAboutAppName("Quoty")
                .withAboutDescription("Beautiful quotes App sharing")
                .withAboutIconShown(true)
                .withAboutVersionShown(true);

        Bundle bundle = new Bundle();
        bundle.putSerializable("data", builder);
        NavHostFragment.findNavController(this).navigate(R.id.action_global_about_libraries, bundle);
    }

    private void checkNotificationEnabled() {
        SharedPreferences sharedPreferences = getPreferenceScreen().getSharedPreferences();
        ListPreference notificationEveryPreference = findPreference(getString(R.string.key_notify_every));
        if (notificationEveryPreference != null) {
            boolean isNotificationEnabled = sharedPreferences.getBoolean(getString(R.string.key_notification), false);
            notificationEveryPreference.setEnabled(isNotificationEnabled);
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        getPreferenceScreen().getSharedPreferences().registerOnSharedPreferenceChangeListener(this);
    }

    @Override
    public void onPause() {
        super.onPause();
        getPreferenceScreen().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);
    }

    private void setTagsEntriesAndValues(MultiSelectListPreference tagsPreference) {
        if (getActivity() == null)return;

        Runnable runnable = () -> {
            QuotesRepository quotesRepository = QuotesRepository.getRepo_instance(getActivity().getApplication());
            String[] tags = quotesRepository.getAllTags();

            getActivity().runOnUiThread(() -> {
                tagsPreference.setEntries(tags);
                tagsPreference.setEntryValues(tags);
                tagsPreference.setEnabled(true);
            });
        };
        AppExecutor.getInstance().getDiskIo().execute(runnable);
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        if (key.equals(getString(R.string.key_lang))
                || key.equals(getString(R.string.key_tags))
                || key.equals(getString(R.string.key_font))
                || key.equals(getString(R.string.key_number_quotes))){

            //get new content according to the new preferences
            AppExecutor.getInstance().getDiskIo().execute(new QuotesFragment.SetNewRowsToShowRunnable(getActivity()));
        }else if (key.equals(getString(R.string.key_notification))
                || key.equals(getString(R.string.key_notify_every))){
            checkNotificationEnabled();
            boolean isNotificationEnabled = sharedPreferences.getBoolean(getString(R.string.key_notification), false);
            if (getContext() != null) AlarmUtil.cancelAlarm(getContext());

            if (isNotificationEnabled) AlarmUtil.startAlarm(getContext());
        }
    }

    @Override
    public boolean onPreferenceChange(Preference preference, Object newValue) {
        if (preference == null || newValue == null) return false;
        if (preference.getKey().equals(getString(R.string.key_number_quotes))){
            try {
                int num = Integer.parseInt(newValue.toString());
                //zero is not allowed
                return num != 0;
            }catch (NumberFormatException e){
                return false;
            }
        }
        return true;
    }
}

File: app/src/main/java/com/hussien/qouty/ui/settings/SettingsFragment.kt
package com.hussien.qouty.ui.settings

import android.app.AlertDialog
import android.content.DialogInterface
import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.os.Bundle
import android.text.InputType
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import androidx.navigation.fragment.findNavController
import androidx.preference.*
import com.hussien.qouty.R
import com.hussien.qouty.ext.capitalizeEveryFirstChar
import com.hussien.qouty.ext.trimAllExtraSpaces
import com.mikepenz.aboutlibraries.LibsBuilder
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch

@AndroidEntryPoint
class SettingsFragment : PreferenceFragmentCompat(), OnSharedPreferenceChangeListener,
    Preference.OnPreferenceChangeListener {

    private val viewModel: SettingsViewModel by viewModels()

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.settings_preference, rootKey)

        lifecycleScope.launchWhenStarted {
            launch { initNumberOfQuotesPref() }
            launch { initTagsPref() }
            launch { initNotifyEveryPref() }
            launch { initFontPref() }
            launch { initLanguagePref() }
        }

        initAboutPref()

        initCreditsPref()
    }

    private suspend fun initLanguagePref() {
        viewModel.languages.collect {
            findPreference<MultiSelectListPreference>(getString(R.string.key_lang))?.run {
                val languages = it.joinToString()
                summary = languages
            }
        }
    }

    private suspend fun initFontPref() {
        viewModel.font.collect {
            findPreference<ListPreference>(getString(R.string.key_font))?.run {
                summary = it.toFontName()
            }
        }
    }

    private fun initCreditsPref() {
        findPreference<Preference>(getString(R.string.key_pref_credits))?.run {
            onPreferenceClickListener =
                Preference.OnPreferenceClickListener {
                    showCreditsDialog()
                    true
                }
        }
    }

    private fun initAboutPref() {
        findPreference<Preference>(getString(R.string.key_pref_about))?.run {
            onPreferenceClickListener = Preference.OnPreferenceClickListener {
                navigateToAboutPage()
                true
            }
        }
    }

    private suspend fun initNotifyEveryPref() {
        viewModel.notificationsEnabled.collect {
            findPreference<ListPreference>(getString(R.string.key_notify_every))?.run {
                isEnabled = it
                summary = viewModel.notifyEvery.map { "${it / 60} Hour" }.first()
            }
        }
    }

    private suspend fun initTagsPref() {
        findPreference<MultiSelectListPreference>(getString(R.string.key_tags))?.run {
            val allTags = viewModel.allTags.first()
            isEnabled = false
            entries = allTags
            entryValues = allTags
            isEnabled = true
            viewModel.tags.collect { tags ->
                summary = tags.joinToString(limit = 4)
            }
        }
    }

    private suspend fun initNumberOfQuotesPref() {
        viewModel.numberOfQuotes.collect {
            findPreference<EditTextPreference>(getString(R.string.key_number_quotes))?.run {
                setOnBindEditTextListener { editText ->
                    editText.inputType = InputType.TYPE_CLASS_NUMBER
                }
                onPreferenceChangeListener = this@SettingsFragment
                summary = it.toString()
            }
        }
    }

    private fun showCreditsDialog() {
        val builder = AlertDialog.Builder(activity)
        builder.setTitle("Credits")
        builder.setMessage("Thanks to harbi for the quotes collection https://github.com/harbi/short-quotes\nApp Icon Designed by Freepik")
        builder.setPositiveButton("OK") { dialog: DialogInterface, _: Int -> dialog.dismiss() }
        builder.create().show()
    }

    private fun navigateToAboutPage() {
        val builder = LibsBuilder()
            .withAboutAppName("Quoty")
            .withAboutDescription("Beautiful quotes App sharing")
            .withAboutIconShown(true)
            .withAboutVersionShown(true)
        val bundle = Bundle()
        bundle.putSerializable("data", builder)
        findNavController().navigate(R.id.action_global_about_libraries, bundle)
    }

    private fun String.toFontName() =
        replace("_regular.ttf", "").split("_")
            .joinToString(separator = " ") { it.trimAllExtraSpaces().capitalizeEveryFirstChar() }


    override fun onResume() {
        super.onResume()
        preferenceScreen.sharedPreferences?.registerOnSharedPreferenceChangeListener(this)
    }

    override fun onPause() {
        super.onPause()
        preferenceScreen.sharedPreferences?.unregisterOnSharedPreferenceChangeListener(this)
    }


    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        lifecycleScope.launchWhenStarted {
            when (key) {
                getString(R.string.key_lang) -> viewModel.settingsManager.updateLanguages(
                    sharedPreferences.getStringSet(key, null)
                )
                getString(R.string.key_tags) -> viewModel.settingsManager.updateTags(
                    sharedPreferences.getStringSet(key, null)
                )
                getString(R.string.key_number_quotes) -> viewModel.settingsManager.updateNumberOfQuotes(
                    sharedPreferences.getString(key, null)?.toInt()
                        ?: SettingsDataStoreManager.DEFAULT_NUMBER_OF_QUOTES
                )
                getString(R.string.key_notification) -> viewModel.settingsManager.updateNotification(
                    sharedPreferences.getBoolean(
                        key,
                        SettingsDataStoreManager.DEFAULT_NOTIFICATIONS_ENABLED
                    )
                )
                getString(R.string.key_notify_every) -> viewModel.settingsManager.updateNotifyEvery(
                    sharedPreferences.getString(
                        key,
                        null
                    )?.toInt() ?: SettingsDataStoreManager.DEFAULT_NOTIFY_EVERY_MINUTES
                )
                getString(R.string.key_font) -> viewModel.settingsManager.updateFont(
                    sharedPreferences.getString(
                        key,
                        null
                    )
                )
            }
        }
    }

    override fun onPreferenceChange(preference: Preference, newValue: Any?): Boolean {
        if (newValue == null) return false
        return if (preference.key == getString(R.string.key_number_quotes)) {
            try {
                val num = newValue.toString().toInt()
                //zero is not allowed
                num != 0
            } catch (e: NumberFormatException) {
                false
            }
        } else true
    }
}